### PR TITLE
ZMQ: Add decodedtx topic for JSON tx publishing

### PR DESCRIPTION
**Pull request by astanway**
- Adds a new ZMQ topic called `decodedtx` that uses `TxToUniv` to decode all newly arrived transaction into JSON and publish them over ZMQ.

- Includes functional tests

### CODE REVIEW

File: src/zmq/zmqpublishnotifier.cpp [hunk 1]
```diff
@@ -195,3 +199,14 @@ bool CZMQPublishRawTransactionNotifier::NotifyTransaction(const CTransaction &tr
...
+bool CZMQPublishDecodedTransactionNotifier::NotifyTransaction(const CTransaction &transaction)
+{
+    uint256 hash = transaction.GetHash();
+    LogPrint(BCLog::ZMQ, "zmq: Publish decodedtx %s\n", hash.GetHex());
+    UniValue result(UniValue::VOBJ);
+    TxToUniv(transaction, uint256(), result, false);
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION | RPCSerializationFlags());
+    ss << result.write();
+    return SendMessage(MSG_DECODEDTX, &(*ss.begin()), ss.size());
+}
```
Comment (promag) [inline src/zmq/zmqpublishnotifier.cpp]: Add newline.

File: test/functional/zmq_test.py [hunk 2]
```diff
@@ -88,6 +90,9 @@ def _zmq_test(self):
         self.sync_all()
 
         for x in range(num_blocks):
+            # Should receive the json broadcasted raw transaction.
```
Comment (promag) [inline test/functional/zmq_test.py]: Remove raw? Same below.

File: src/zmq/zmqpublishnotifier.cpp [hunk 3]
```diff
@@ -195,3 +199,14 @@ bool CZMQPublishRawTransactionNotifier::NotifyTransaction(const CTransaction &tr
     ss << transaction;
     return SendMessage(MSG_RAWTX, &(*ss.begin()), ss.size());
 }
+
+bool CZMQPublishDecodedTransactionNotifier::NotifyTransaction(const CTransaction &transaction)
+{
+    uint256 hash = transaction.GetHash();
+    LogPrint(BCLog::ZMQ, "zmq: Publish decodedtx %s\n", hash.GetHex());
+    UniValue result(UniValue::VOBJ);
+    TxToUniv(transaction, uint256(), result, false);
```
Comment (promag) [inline src/zmq/zmqpublishnotifier.cpp]: Have you tried `TxToJSON`?

File: src/zmq/zmqpublishnotifier.cpp [hunk 4]
```diff
@@ -9,13 +9,17 @@
 #include "validation.h"
 #include "util.h"
 #include "rpc/server.h"
+#include "rpc/protocol.h"
```
Comment (promag) [inline src/zmq/zmqpublishnotifier.cpp]: Nit, sort includes.

File: src/zmq/zmqpublishnotifier.cpp [hunk 5]
```diff
@@ -9,13 +9,17 @@
...
 
-static const char *MSG_HASHBLOCK = "hashblock";
-static const char *MSG_HASHTX    = "hashtx";
-static const char *MSG_RAWBLOCK  = "rawblock";
-static const char *MSG_RAWTX     = "rawtx";
+static const char *MSG_HASHBLOCK     = "hashblock";
+static const char *MSG_HASHTX        = "hashtx";
+static const char *MSG_RAWBLOCK      = "rawblock";
+static const char *MSG_RAWTX         = "rawtx";
+static const char *MSG_DECODEDTX     = "decodedtx";
```
Comment (promag) [inline src/zmq/zmqpublishnotifier.cpp]: Nit, sort.

File: test/functional/zmq_test.py [hunk 6]
```diff
@@ -109,6 +114,10 @@ def _zmq_test(self):
         payment_txid = self.nodes[1].sendtoaddress(self.nodes[0].getnewaddress(), 1.0)
         self.sync_all()
 
+        # Should receive the json broadcasted raw transaction.
+        decodedtx = self.decodedtx.receive()
+        assert_equal(payment_txid, json.loads(decodedtx[3:])['txid'])
```
Comment (promag) [inline test/functional/zmq_test.py]: Care to explain the offset?

**Review by promag — COMMENTED** [review]
Concept ACK and welcome!

File: src/zmq/zmqpublishnotifier.cpp (hunk 3 above)
Comment (astanway) [reply src/zmq/zmqpublishnotifier.cpp]: `TxToJSON` just calls `TxToUniv` internally, and also adds that extra `hashBlock` check, which is not strictly necessary imo. If we think that extra info would be useful though (confirmations, time, blocktime, blockhash), we can try to get the block hash beforehand and switch to using this method.

File: test/functional/zmq_test.py (hunk 6 above)
Comment (astanway) [reply test/functional/zmq_test.py]: Looks like it was some kind of header added by `CDataStream` - I'm not entirely sure, tbh. I switched it over to a vanilla string and char array and the offset went away.

**Comment by jonasschnelli** [comment]
Tend to NACK.
One of the downsides of our RPC concept is that everything is based on JSON. JSON has serval downsides (parsing is in general a minefield).

IMO ZMQ should be JSON-free and ideally binary/net-serialized only.

Why not just listening via ZMQ and fetch the required data via RPC if a ZMQ notification arrives?

Or where is the benefit of having a decoded JSON rather then decoding it on the client side?

**Comment by astanway** [comment]
The benefit is much greater speed than through RPC, primarily because ZMQ can be configured to use a unix domain socket instead of TCP (which RPC currently does not support). In addition, to listen for `txid` via ZMQ and then make an additional RPC request with it seems like a waste, when the needed data could be sent right off the bat.

**Comment by dcousens** [comment]
@astanway perhaps a better long term outcome is some kind of IPC socket instead of the RPC/ZMQ.

**Comment by promag** [comment]
>  as opposed to rolling your own decoder which might differ / be buggy / lack features, especially as time goes on and the RPC output is possibly changed

There are libraries for this already.

IMO rich notifications are fine, even though it is possible to have the same data thru RPC or some client library. With `rawtx` you just have to parse the JSON and for prototypes or simple applications it could be that simple. I would just add a note regarding performance vs `rawtx`.

**Comment by laanwj** [comment]
> IMO ZMQ should be JSON-free and ideally binary/net-serialized only.

I tend to agree here. The point of ZMQ is that it's fast, and adding JSON overhead (formatting on the server, parsing on the client) negates that. 
Why doesn't your application process the raw transaction data instead?

> as opposed to rolling your own decoder which might differ / be buggy / lack features, especially as time goes on and the RPC output is possibly changed.

I think you have things reversed there - the chance of JSON output changing under you is much larger than the binary encoding of transactions changing, as that would be a consensus change, which is by necessarily advertised widely. Processing binary transaction data directly increases robustness of whatever you're building against upstream changes.

**Comment by promag** [comment]
FYI this was discussed here https://github.com/bitcoin/bitcoin/pull/6103#issuecomment-99076113.

This being an option and using the same RPC format I keep the concept ACK. IMO the overhead is not that much (or it shouldn't be). Also, everything is in place to easily add new notifiers.

**Comment by astanway** [comment]
As far as performance goes - if the option is not enabled, it has zero impact (as far as I can understand from the rest of ZMQ project). I haven't really heard a compelling reason to _not_ add it - I think it's a very convenient and useful feature and will enable a lot of people to start developing quickly, without needing to write their own transaction parser from scratch (difficult) or using RPC (slow).

**Comment by dcousens** [comment]
>  transaction parser from scratch (difficult) 

There are libraries in nearly every language you would use.  It is probably a 1-line function call by comparison.
The reasonable advantage of the RPC is the *extra* information encoded into the JSON format RE block information,  that would otherwise be an *extra* round trip.

**Comment by astanway** [comment]
> There are libraries in nearly every language you would use. It is probably a 1-line function call.

Truly there are not; the state of transaction parsers is in pretty horrible shape and none really match the canonical RPC JSON output.

> The only reasonable advantage of the RPC 
Sorry, not sure what we're talking about - I am defending using this ZMQ extension, not RPC.

**Comment by dcousens** [comment]
> I am defending using this ZMQ extension, not RPC.

Apologies,  I meant the *RPC JSON format*.  Edited my reply to fit that.

**Comment by luke-jr** [comment]
This is too wasteful and pointless IMO. Just decode the raw transaction on the client side.

**Comment by jnewbery** [comment]
I'm a -0 on this, for the reasons from @jonasschnelli and @laanwj .

This doesn't seem to have garnered support. @astanway - are you happy to close this?

**Comment by astanway** [comment]
I still think it can be quite useful to a lot of folks who want to get started developing on top of the blockchain. There's no reason to force parsing on users when we can do a little bit of work to do it natively and in a simple (aka, readily both human and machine readable) format like JSON. However, if ease of application development is not a priority for the project, then sure - happy to close.

**Comment by jnewbery** [comment]
> if ease of application development is not a priority...

Thanks @astanway , this isn't so much about priorities, as whether the concept has gathered support from other contributors. This PR has 5 concept NACKs from regular contributors, so it seems unlikely to get support for merge.

I've had plenty of my PRs NACKed in the past. It's part of the open source process.

**Comment by promag** [comment]
Agree with closing.

> Truly there are not; the state of transaction parsers is in pretty horrible shape and none really match the canonical RPC JSON output.

Perhaps you could help to improve this instead.

**Pull request closed without merging**